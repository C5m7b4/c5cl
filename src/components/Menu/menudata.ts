export const menudata = [
  {
    title: 'Home',
  },
  {
    title: 'Services',
    submenu: [
      {
        title: 'web design',
      },
      {
        title: 'web development',
        submenu: [
          {
            title: 'frontend',
          },
          {
            title: 'backend',
            submenu: [
              {
                title: 'NodeJS',
              },
              {
                title: 'PHP',
              },
            ],
          },
        ],
      },
      {
        title: 'seo',
      },
    ],
  },
  {
    title: 'About',
  },
];
