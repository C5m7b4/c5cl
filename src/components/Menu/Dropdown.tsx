import React from 'react';
import MenuItems from './MenuItems';

interface SubMenuProps {
  title: string;
}

interface DropDownProps {
  submenus: SubMenuProps[];
  dropdown: boolean;
  depthLevel: number;
}

const Dropdown = (props: DropDownProps) => {
  let { depthLevel } = props;
  depthLevel = depthLevel + 1;
  const dropdownClass = depthLevel > 1 ? 'dropdown-submenu' : '';

  return (
    <ul className={`dropdown ${dropdownClass} ${props.dropdown ? 'show' : ''}`}>
      {props.submenus.map((submenu, index) => (
        <li key={`sbm-${index}`} className="menu-items">
          <MenuItems
            items={submenu}
            key={`mi-sb-${index}`}
            depthLevel={depthLevel}
          />
        </li>
      ))}
    </ul>
  );
};

export default Dropdown;
