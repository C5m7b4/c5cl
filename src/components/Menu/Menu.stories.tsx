import React from 'react';
import { Story, Meta } from '@storybook/react';

import Menu, { MenuProps } from './Menu';
import { menudata } from './menudata';

export default {
  title: 'Widgets/Menu',
  component: Menu,
} as Meta;

const Template: Story<MenuProps> = (args) => <Menu {...args} />;

export const MenuExample = Template.bind({});
MenuExample.args = {
  logo: 'Mikto',
  items: menudata,
};
