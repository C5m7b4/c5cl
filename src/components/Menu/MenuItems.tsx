import React, { useState, useEffect, useRef } from 'react';
import Dropdown from './Dropdown';

interface SubMenuProps {
  title: string;
}

interface ItemProps {
  title: string;
  submenu?: SubMenuProps[];
}

interface MenuItemsProps {
  items: ItemProps;
  depthLevel: number;
}

const MenuItems = (props: MenuItemsProps) => {
  const [dropdown, setDropdown] = useState(false);

  let ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    const handler = (event: any) => {
      // @ts-ignore
      if (dropdown && ref.current && !ref.current.contains(event.target)) {
        setDropdown(false);
      }
    };

    document.addEventListener('mousedown', handler);
    document.addEventListener('touchstart', handler);

    return () => {
      document.removeEventListener('mousedown', handler);
      document.removeEventListener('touchstart', handler);
    };
  }, [dropdown]);

  const onMouseEnter = (e: React.MouseEvent) => {
    window.innerWidth > 960 && setDropdown(true);
  };

  const onMouseLeave = (e: React.MouseEvent) => {
    window.innerWidth > 960 && setDropdown(false);
  };

  return (
    <div
      className="menu-items"
      ref={ref}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      {props.items.submenu ? (
        <>
          <button
            type="button"
            aria-haspopup="menu"
            aria-expanded={dropdown ? 'true' : 'false'}
            onClick={() => setDropdown((prev) => !prev)}
          >
            {props.items.title}{' '}
            {props.depthLevel > 0 ? (
              <span>&raquo;</span>
            ) : (
              <span className="arrow" />
            )}
          </button>
          <Dropdown
            submenus={props.items.submenu}
            dropdown={dropdown}
            depthLevel={props.depthLevel}
          />
        </>
      ) : (
        <a href="/#">{props.items.title}</a>
      )}
    </div>
  );
};

export default MenuItems;
