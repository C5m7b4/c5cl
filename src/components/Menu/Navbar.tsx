import React from 'react';
import MenuItems from './MenuItems';

interface SubMenuProps {
  title: string;
}

interface ItemProps {
  title: string;
  submenu?: SubMenuProps[];
}

interface NavbarProps {
  items: ItemProps[];
}

const Navbar = (props: NavbarProps) => {
  const depthLevel = 0;
  return (
    <nav>
      <ul className="menus">
        {props.items.map((menu, index) => {
          return (
            <MenuItems
              items={menu}
              key={`mi-${index}`}
              depthLevel={depthLevel}
            />
          );
        })}
      </ul>
    </nav>
  );
};

export default Navbar;
