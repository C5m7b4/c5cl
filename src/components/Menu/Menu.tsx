import React from 'react';
import Navbar from './Navbar';
import './Menu.css';

interface SubMenuProps {
  title: string;
}

interface ItemProps {
  title: string;
  submenu?: SubMenuProps[];
}

export interface MenuProps {
  logo?: any;
  items: ItemProps[];
}

const Menu = (props: MenuProps) => {
  return (
    <header>
      <div className="nav-area">
        <a href="/#" className="logo">
          {props.logo}
        </a>
        <Navbar items={props.items} />
      </div>
    </header>
  );
};

export default Menu;
