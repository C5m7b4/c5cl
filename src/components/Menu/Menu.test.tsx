import React from 'react';
import { render, screen, fireEvent, getByRole } from '@testing-library/react';

import { menudata } from './menudata';

import Menu from './Menu';

const Sandbox = () => {
  return (
    <div>
      <Menu items={menudata} />
      <div className="other">
        Here is some other div for clicking outside of our component
      </div>
    </div>
  );
};

describe('Menu', () => {
  test('should render correctly', () => {
    const { container } = render(<Menu items={menudata} />);
    expect(container).toMatchSnapshot();
  });
  test('should show submenu', () => {
    const { container, getByRole } = render(<Menu items={menudata} />);

    const btn = screen.getByRole('button', { name: 'Services' });
    fireEvent.mouseOver(btn);
    const menuItem = screen.getByRole('button', { name: 'web development Â»' });
    fireEvent.mouseOver(menuItem);
    fireEvent.mouseLeave(menuItem);
    expect(container).toMatchSnapshot();
    fireEvent.mouseLeave(btn);

    fireEvent.click(btn);
  });
  test('should handle click outside', () => {
    const { container, getByRole } = render(<Sandbox />);
    const btn = screen.getByRole('button', { name: 'Services' });
    fireEvent.mouseOver(btn);

    const body = container.querySelector('.other') as HTMLDivElement;
    fireEvent.mouseDown(body);
  });
});
