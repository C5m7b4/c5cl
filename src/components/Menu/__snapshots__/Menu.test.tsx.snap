// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu should render correctly 1`] = `
<div>
  <header>
    <div
      class="nav-area"
    >
      <a
        class="logo"
        href="/#"
      />
      <nav>
        <ul
          class="menus"
        >
          <div
            class="menu-items"
          >
            <a
              href="/#"
            >
              Home
            </a>
          </div>
          <div
            class="menu-items"
          >
            <button
              aria-expanded="false"
              aria-haspopup="menu"
              type="button"
            >
              Services
               
              <span
                class="arrow"
              />
            </button>
            <ul
              class="dropdown  "
            >
              <li
                class="menu-items"
              >
                <div
                  class="menu-items"
                >
                  <a
                    href="/#"
                  >
                    web design
                  </a>
                </div>
              </li>
              <li
                class="menu-items"
              >
                <div
                  class="menu-items"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="menu"
                    type="button"
                  >
                    web development
                     
                    <span>
                      »
                    </span>
                  </button>
                  <ul
                    class="dropdown dropdown-submenu "
                  >
                    <li
                      class="menu-items"
                    >
                      <div
                        class="menu-items"
                      >
                        <a
                          href="/#"
                        >
                          frontend
                        </a>
                      </div>
                    </li>
                    <li
                      class="menu-items"
                    >
                      <div
                        class="menu-items"
                      >
                        <button
                          aria-expanded="false"
                          aria-haspopup="menu"
                          type="button"
                        >
                          backend
                           
                          <span>
                            »
                          </span>
                        </button>
                        <ul
                          class="dropdown dropdown-submenu "
                        >
                          <li
                            class="menu-items"
                          >
                            <div
                              class="menu-items"
                            >
                              <a
                                href="/#"
                              >
                                NodeJS
                              </a>
                            </div>
                          </li>
                          <li
                            class="menu-items"
                          >
                            <div
                              class="menu-items"
                            >
                              <a
                                href="/#"
                              >
                                PHP
                              </a>
                            </div>
                          </li>
                        </ul>
                      </div>
                    </li>
                  </ul>
                </div>
              </li>
              <li
                class="menu-items"
              >
                <div
                  class="menu-items"
                >
                  <a
                    href="/#"
                  >
                    seo
                  </a>
                </div>
              </li>
            </ul>
          </div>
          <div
            class="menu-items"
          >
            <a
              href="/#"
            >
              About
            </a>
          </div>
        </ul>
      </nav>
    </div>
  </header>
</div>
`;

exports[`Menu should show submenu 1`] = `
<div>
  <header>
    <div
      class="nav-area"
    >
      <a
        class="logo"
        href="/#"
      />
      <nav>
        <ul
          class="menus"
        >
          <div
            class="menu-items"
          >
            <a
              href="/#"
            >
              Home
            </a>
          </div>
          <div
            class="menu-items"
          >
            <button
              aria-expanded="false"
              aria-haspopup="menu"
              type="button"
            >
              Services
               
              <span
                class="arrow"
              />
            </button>
            <ul
              class="dropdown  "
            >
              <li
                class="menu-items"
              >
                <div
                  class="menu-items"
                >
                  <a
                    href="/#"
                  >
                    web design
                  </a>
                </div>
              </li>
              <li
                class="menu-items"
              >
                <div
                  class="menu-items"
                >
                  <button
                    aria-expanded="false"
                    aria-haspopup="menu"
                    type="button"
                  >
                    web development
                     
                    <span>
                      »
                    </span>
                  </button>
                  <ul
                    class="dropdown dropdown-submenu "
                  >
                    <li
                      class="menu-items"
                    >
                      <div
                        class="menu-items"
                      >
                        <a
                          href="/#"
                        >
                          frontend
                        </a>
                      </div>
                    </li>
                    <li
                      class="menu-items"
                    >
                      <div
                        class="menu-items"
                      >
                        <button
                          aria-expanded="false"
                          aria-haspopup="menu"
                          type="button"
                        >
                          backend
                           
                          <span>
                            »
                          </span>
                        </button>
                        <ul
                          class="dropdown dropdown-submenu "
                        >
                          <li
                            class="menu-items"
                          >
                            <div
                              class="menu-items"
                            >
                              <a
                                href="/#"
                              >
                                NodeJS
                              </a>
                            </div>
                          </li>
                          <li
                            class="menu-items"
                          >
                            <div
                              class="menu-items"
                            >
                              <a
                                href="/#"
                              >
                                PHP
                              </a>
                            </div>
                          </li>
                        </ul>
                      </div>
                    </li>
                  </ul>
                </div>
              </li>
              <li
                class="menu-items"
              >
                <div
                  class="menu-items"
                >
                  <a
                    href="/#"
                  >
                    seo
                  </a>
                </div>
              </li>
            </ul>
          </div>
          <div
            class="menu-items"
          >
            <a
              href="/#"
            >
              About
            </a>
          </div>
        </ul>
      </nav>
    </div>
  </header>
</div>
`;
